
Inspecting vendor Android DT for x96-q with H313 we wee different than in H616 CPU Opp.
Add dedicated CPU Opp for H313 and switch x96-q to use them instead using default H616 Opp.

Developed by piotr.oniszczuk@gmail.com

diff --speed-large-files --no-dereference --minimal -Naur linux-6.2.9/arch/arm64/boot/dts/allwinner/sun50i-h313-x96-q.dts linux-6.2.9/arch/arm64/boot/dts/allwinner/sun50i-h313-x96-q.dts
--- linux-6.2.9/arch/arm64/boot/dts/allwinner/sun50i-h313-x96-q.dts	2023-03-31 18:25:20.884219530 +0200
+++ linux-6.2.9/arch/arm64/boot/dts/allwinner/sun50i-h313-x96-q.dts	2023-03-31 18:24:22.114221667 +0200
@@ -87,6 +88,25 @@
 
 &cpu0 {
 	cpu-supply = <&reg_dcdca>;
+	operating-points-v2 = <&cpu_opp_table_h313>;
+	status = "okay";
+};
+
+&cpu1 {
+	cpu-supply = <&reg_dcdca>;
+	operating-points-v2 = <&cpu_opp_table_h313>;
+	status = "okay";
+};
+
+&cpu2 {
+	cpu-supply = <&reg_dcdca>;
+	operating-points-v2 = <&cpu_opp_table_h313>;
+	status = "okay";
+};
+
+&cpu3 {
+	cpu-supply = <&reg_dcdca>;
+	operating-points-v2 = <&cpu_opp_table_h313>;
 	status = "okay";
 };
 
diff --speed-large-files --no-dereference --minimal -Naur linux-6.2.9/arch/arm64/boot/dts/allwinner/sun50i-h616-cpu-opp.dtsi linux-6.2.9/arch/arm64/boot/dts/allwinner/sun50i-h616-cpu-opp.dtsi
--- linux-6.2.9/arch/arm64/boot/dts/allwinner/sun50i-h616-cpu-opp.dtsi	2023-03-31 18:25:20.654219539 +0200
+++ linux-6.2.9/arch/arm64/boot/dts/allwinner/sun50i-h616-cpu-opp.dtsi	2023-03-31 18:06:05.530928422 +0200
@@ -83,5 +83,69 @@
 //			opp-microvolt = <1120000 1120000 1120000>;
 //		};
 	};
-};
 
+/*
+	X96-q DDR3 vendor Android DT:
+	480000000	900mV
+	600000000	900mV
+	792000000	900mV
+	1008000000	920mV
+	1200000000	980mV
+	1344000000	1120mV
+	1416000000	1140mV
+	1512000000	1160mV
+*/
+
+	cpu_opp_table_h313: cpu-opp-table-h313 {
+		compatible = "operating-points-v2";
+		opp-shared;
+
+		opp-480000000 {
+			clock-latency-ns = <244144>; /* 8 32k periods */
+			opp-hz = /bits/ 64 <480000000>;
+			opp-microvolt = <900000 900000 900000>;
+		};
+
+		opp-600000000 {
+			clock-latency-ns = <244144>; /* 8 32k periods */
+			opp-hz = /bits/ 64 <600000000>;
+			opp-microvolt = <900000 900000 900000>;
+		};
+
+		opp-792000000 {
+			clock-latency-ns = <244144>; /* 8 32k periods */
+			opp-hz = /bits/ 64 <792000000>;
+			opp-microvolt = <900000 900000 900000>;
+		};
+
+		opp-1008000000 {
+			clock-latency-ns = <244144>; /* 8 32k periods */
+			opp-hz = /bits/ 64 <1008000000>;
+			opp-microvolt = <920000 920000 920000>;
+		};
+
+		opp-1200000000 {
+			clock-latency-ns = <244144>; /* 8 32k periods */
+			opp-hz = /bits/ 64 <1200000000>;
+			opp-microvolt = <980000 980000 980000>;
+		};
+
+		opp-1344000000 {
+			clock-latency-ns = <244144>; /* 8 32k periods */
+			opp-hz = /bits/ 64 <1344000000>;
+			opp-microvolt = <1120000 1120000 1120000>;
+		};
+
+		opp-1416000000 {
+			clock-latency-ns = <244144>; /* 8 32k periods */
+			opp-hz = /bits/ 64 <1416000000>;
+			opp-microvolt = <1150000 1150000 1150000>;
+		};
+
+		opp-1512000000 {
+			clock-latency-ns = <244144>; /* 8 32k periods */
+			opp-hz = /bits/ 64 <1512000000>;
+			opp-microvolt = <1160000 1160000 1160000>;
+		};
+	};
+};
